### ISSUE TITLE
Shuffle buffer causes OOM on CPU

### ISSUE THREAD

**Issue by stefan-falk** [tensorflow/tensor2tensor#1210]
I noticed that with 1.10.0 a shuffle buffer get build up before training:

```
2018-11-09 11:48:04.525172: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:97] Filling up shuffle buffer (this may take a while): 391 of 512
2018-11-09 11:48:14.233178: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:97] Filling up shuffle buffer (this may take a while): 396 of 512
2018-11-09 11:48:29.700824: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:97] Filling up shuffle buffer (this may take a while): 400 of 512
2018-11-09 11:48:33.617605: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:97] Filling up shuffle buffer (this may take a while): 402 of 512
2018-11-09 11:48:50.017594: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:97] Filling up shuffle buffer (this may take a while): 406 of 512
2018-11-09 11:48:56.350018: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:97] Filling up shuffle buffer (this may take a while): 407 of 512
```

However, for one of my larger t2t-problems this seems to cause an OOM error (CPU RAM). I am not sure if this operation happened before 1.10.0 but in any case I'd like to do something against this OOM error. 

Why is there a shuffle buffer getting build up and can I disable it or at least control its size s.t. it fits into memory?

**Comment by stefan-falk** [comment]
I noticed that `--hp_batch_size=<value>` or `--hparams='batch_size=<value>'` will trigger the creation of a shuffle buffer.

**Comment by cbockman** [comment]
Not on t2t team, but I'm guessing you're hitting the internal shuffling:

https://github.com/tensorflow/tensor2tensor/blob/master/tensor2tensor/utils/data_reader.py#L155
https://github.com/tensorflow/tensor2tensor/blob/8bcbdccf85c0fc60f07945c469ff3213d2e0810d/tensor2tensor/data_generators/problem.py#L966
https://github.com/tensorflow/tensor2tensor/blob/8bcbdccf85c0fc60f07945c469ff3213d2e0810d/tensor2tensor/data_generators/problem.py#L559

> Why is there a shuffle buffer getting build up and can I disable it or at least control its size s.t. it fits into memory?

One or more of those links should give you the tools to disable.

That said, it is there because otherwise you end up with data being cycled through in a deterministic fashion, which is generally going to be subpar to shuffled data, 1) because you'll be running through the data in the same order and 2) your original data may have hidden deterministic ordering.

If you really don't want to / can't do data set shuffling, then you could instead shard the data to a very high number of shards, since (iirc) t2t will grab the shards in random order.

**Comment by stefan-falk** [comment]
@cbockman Thanks for answering!

I've already tried to set `shuffle_buffer_size` though (e.g. `--hp_shuffle_buffer_size`). For some reason it seems that there are two buffers build up. On that uses size 1024 per default and another that users 512 per default. I was not able to find the place in the code which would set that.

**Comment by cbockman** [comment]
> On that uses size 1024 per default and another that users 512 per default. I was not able to find the place in the code which would set that.

I linked you this above, second link I provided is where the 512 batch is used. You can ctrl-f to trace it back to
https://github.com/tensorflow/tensor2tensor/blob/8bcbdccf85c0fc60f07945c469ff3213d2e0810d/tensor2tensor/data_generators/problem.py#L806, where it is specifically set.

**Comment by stefan-falk** [comment]
@cbockman Ah, thank you I didn't realize. And now I also see that this behavior is new and actually came with 1.10.0!

**Comment by JanithT-Lboro** [comment]
Hello, @stefan-falk 
Did you ever find a solution to your problem as I am experiencing the same issue currently?
